spec/models/document_spec.rb

```diff
@@ -39,4 +45,216 @@ def self.title
+          field3: %w(x y z),
+          document_type: "my_document_type",
+          bulk_published: true,
+        }
+      })
+  }
+  let(:document) { MyDocumentType.from_publishing_api(payload) }
+
+  before do
+    allow_any_instance_of(FinderSchema).to receive(:load_schema_for).with("my_document_types").
+      and_return(finder_schema)
+  end
+
+  context "successful #publish!" do
+    before do
+      stub_publishing_api_publish(document.content_id, {})
+      stub_any_rummager_post_with_queueing_enabled
+      @email_alert_api = email_alert_api_accepts_alert
+    end
+
+    it "sends a payload to Publishing API" do
+      expect(document.publish!).to eq(true)
+
+      assert_publishing_api_publish(document.content_id)
+    end
+
+    it "sends a payload to Rummager" do
+      expect(document.publish!).to eq(true)
+
+      assert_rummager_posted_item(
+        "title" => "Example document",
+        "description" => "This is a summary",
+        "indexable_content" => "Header " + (["This is the long body of an example document"] * 10).join(" "),
+        "link" => "/my-document-types/example-document",
+        "public_timestamp" => "2015-11-16T11:53:30+00:00",
+        "field1" => "2015-12-01",
+        "field2" => "open",
+      )
+    end
+
+    it "alerts the email API for major updates" do
+      document.update_type = "major"
+
+      expect(document.publish!).to eq(true)
+
+      assert_email_alert_sent(
+        "tags" => {
+          "format" => "my_document_type",
+          "field1" => "2015-12-01",
+          "field2" => "open",
+          "field3" => %w(x y z),
+        },
+        "document_type" => "my_document_type"
+      )
+    end
+
+    it "doesn't alerts the email API for minor updates" do
+      document.update_type = "minor"
+
+      expect(document.publish!).to eq(true)
+
+      expect(@email_alert_api).to_not have_been_requested
+    end
+  end
+
+  context "unsuccessful #publish!" do
+    it "notifies Airbrake and returns false if publishing-api does not return status 200" do
+      expect(Airbrake).to receive(:notify)
+      stub_publishing_api_publish(document.content_id, {}, status: 503)
+      stub_any_rummager_post_with_queueing_enabled
+      expect(document.publish!).to eq(false)
+    end
+
+    it "notifies Airbrake and returns false if rummager does not return status 200" do
+      expect(Airbrake).to receive(:notify)
+      stub_publishing_api_publish(document.content_id, {})
+      stub_request(:post, %r{#{Plek.new.find('search')}/documents}).to_return(status: 503)
+      expect(document.publish!).to eq(false)
+    end
+  end
+
+  describe "#save!" do
+    before do
+      publishing_api_has_item(payload)
+      Timecop.freeze(Time.parse("2015-12-18 10:12:26 UTC"))
+    end
+
+    it "saves document" do
+      stub_any_publishing_api_put_content
+      stub_any_publishing_api_patch_links
+
+      c = MyDocumentType.find(payload["content_id"])
+      expect(c.save!).to eq(true)
```

Comment:
I agree – `save!` doesn't behave correctly when there's a network issue. I have a fix for it locally but I'll raise a separate pull request because the fix requires a change to the production code, whereas I'd prefer to keep this as a test-only change.